### DISCUSSION TITLE
Get all green variations inside image

### DISCUSSION THREAD

**Discussion by magebot1** [ImageMagick/ImageMagick#3639] [Help] [answered]
How i could get all green variations in this image?
![9440](https://user-images.githubusercontent.com/73500606/117346176-c984ca00-ae7d-11eb-8a04-72bcf1c0cb53.png)

More samples:

![3408](https://user-images.githubusercontent.com/73500606/117346183-d0134180-ae7d-11eb-9d48-8fd5ea067cc5.png)

![5710](https://user-images.githubusercontent.com/73500606/117346187-d0abd800-ae7d-11eb-8ecd-d7efbb3627be.png)

![622](https://user-images.githubusercontent.com/73500606/117346188-d1446e80-ae7d-11eb-89d7-b25c1bb7421b.png)

![7934](https://user-images.githubusercontent.com/73500606/117346189-d1446e80-ae7d-11eb-81aa-e8ca5f5891b9.png)

**Comment by snibgo** [comment]
What version of IM, on what platform? I will assume IM v7 on Windows BAT.

First, define what you mean by "green variations". One possible definition is "pixels where G>R and G>B".

Naming your first input file as "greens.png":
```
magick ^
  greens.png ^
  -unique-colors +write info: ^
  -fx "(u.r>=u.g)?0:u" ^
  -fx "(u.b>=u.g)?0:u" ^
  -unique-colors +write info: ^
  g.png
```
The output, g.png, has all the "green variations", plus one black pixel.

It also creates text output:
```
greens.png PNG 5006x1 5006x1+0+0 8-bit sRGB 17889B 0.000u 0:00.005
greens.png PNG 1310x1 1310x1+0+0 8-bit sRGB 17889B 0.000u 0:00.001
```
This shows that your input has 5006 unique colours, of which 1310 are "green variations" plus one for the black.

**Reply by magebot1** [reply]
Hi snibgo, thank you for your time, is possible to save to a text file all colors inside that "g.png" file in hex?
Something like
`#3A8E24,#3A8E25,#3B8E25,3B8F26,#3C9026....etc`

**Comment by snibgo** [comment]
You can write the image to "txt:". For example, showing the first few lines of the output:
```
magick g.png txt:

# ImageMagick pixel enumeration: 1310,1,65535,srgba
0,0: (0,0,0,0)  #00000000  none
1,0: (8738,14135,3341,65535)  #22370DFF  srgba(34,55,13,1)
2,0: (9766,14649,6168,65535)  #263918FF  srgba(38,57,24,1)
3,0: (12079,15163,7453,65535)  #2F3B1DFF  srgba(47,59,29,1)
4,0: (10537,15934,7453,65535)  #293E1DFF  srgba(41,62,29,1)
5,0: (12079,14906,9252,65535)  #2F3A24FF  srgba(47,58,36,1)
```
With a shell script, you can separate the middle field.

**Reply by magebot1** [reply]
Thanks, im trying to understand how you specified the "green" color, how would be if I would like to get the "blue" variations 
![image](https://user-images.githubusercontent.com/73500606/117362012-2342bf80-ae91-11eb-90cb-5c1556a89ac6.png)

**Comment by snibgo** [comment, marked as answer]
My command includes the lines:
```
  -fx "(u.r>=u.g)?0:u" ^
  -fx "(u.b>=u.g)?0:u" ^
```
The first "-fx" changes pixel values to either 0 (zero, which is black), or "u" which is no change. If the red channel is greater than or equal to the green channel, the value is changed to zero. Otherwise, it makes no change.

Similarly the next line: if the blue channel is greater than or equal to the green channel, the value is changed to zero.

So, if you want all the "blue variations", you can change those lines to:
```
  -fx "(u.r>=u.b)?0:u" ^
  -fx "(u.g>=u.b)?0:u" ^
```

**Comment by fmw42** [comment]
Here is another way (Unix Syntax).  I convert the image to HSV colorspace. Then separate the hue channel and threshold it on green color and save that as a mask. I then put the mask into the alpha channel of the image to make the non-green colors transparent. Then I get the unique colors and use sparse-color: to get the rgba values only for the non-fully-transparent pixels. I then strip off the coordinates.

Input:

![samples](https://user-images.githubusercontent.com/13458218/117370330-17e49980-ae7b-11eb-9684-3d2081a80b72.png)

```
color=120  # green hue
toler=20
magick samples.png -alpha off \
\( +clone -colorspace HSV -channel 0 -separate +channel \
-color-threshold "gray(%[fx:($color-$toler)*255/360])-gray(%[fx:($color+$toler)*255/360])" +write mask.png \) \
-alpha off -compose copy_opacity -composite \
-unique-colors sparse-color: | tr " " "\n" | sed -n 's/^.*\(srgb.*\)$/\1/p' > colors.txt

```

Mask image:

![mask](https://user-images.githubusercontent.com/13458218/117370238-f5eb1700-ae7a-11eb-89cf-2428136624ac.png)

Text file of rgba  values:

[colors.txt](https://github.com/ImageMagick/ImageMagick/files/6437815/colors.txt)

**Alternate:**

To get text file hex values:

```

color=120  # green hue
toler=20
magick samples.png -alpha off \
\( +clone -colorspace HSV -channel 0 -separate +channel \
-color-threshold "gray(%[fx:($color-$toler)*255/360])-gray(%[fx:($color+$toler)*255/360])" +write mask.png \) \
-compose multiply -composite \
-unique-colors -alpha off txt: | tail -n +2 | sed -n 's/^.*\(\#[^ ]*\).*$/\1/p' > colors2.txt

```

Text file of hex values:

[colors2.txt](https://github.com/ImageMagick/ImageMagick/files/6437856/colors2.txt)

_But it will also include one value of black._
